src/core/lwt_stream.ml

```diff
@@ -130,28 +132,31 @@ let clone s =
      | _ -> ());
   {
     source = s.source;
+    closed = s.closed;
     node = s.node;
     last = s.last;
     hooks = s.hooks;
   }
 
-let from f =
+let from_source source =
   let last = new_node () in
-  {
-    source = From { from_create = f; from_thread = Lwt.return_unit };
-    node = last;
-    last = ref last;
-    hooks = ref [];
+  let closed, closed_w = Lwt.wait () in
+  let mark_closed () =
+    if not (Lwt.is_sleeping closed) then Lwt.wakeup closed_w () in
+  { source = source
+  ; closed = closed
+  ; node = last
+  ; last = ref last
+  ; hooks = ref [mark_closed]
   }
 
+let from f =
+  from_source (From { from_create = f; from_thread = Lwt.return_unit })
+
 let from_direct f =
-  let last = new_node () in
-  {
-    source = From_direct f;
-    node = last;
-    last = ref last;
-    hooks = ref [];
-  }
+  let t = from_source (From_direct f) in
+  List.iter (fun f -> f ()) !(t.hooks);
```

Comment:
In the case of a direct source, the stream is terminated upon creation. While the user doesn't have an opportunity to register any termination callbacks, the module itself may have via the `from_source` constructor. And indeed it currently does do just that to handle the `is_closed` functionality that this pull request adds (your comment above).